engineering/conventions-go.md

```diff
@@ -0,0 +1,62 @@
+
+# Go Development Conventions
+
+## Introduction
+
+This guide documents development conventions for Go at source{d}. Check [general conventions](conventions.md) for language-independent conventions that are also applicable for Go projects.
+
+## Supported Go Versions
+
+* Support latest two stable major versions of Go (e.g. 1.9.x and 1.10.x). Both versions should be included in Travis CI.
+  * Some of our projects may support more versions, specially when there is wide-adoption outside source{d} (e.g. [go-git](https://github.com/src-d/go-git) currently supports three).
+  * We try to support new Go versions as fast as possible. Versions to test in Travis CI include `tip` to help us identifying problems with next version ahead of time.
+
+## Dependency Management
+
+* If your project is an application:
+  * Use [dep](https://github.com/golang/dep) to manage dependencies.
+  * Check in the `vendor` directory into git.
+* If your project is a library:
+  * Use [gopkg.in](http://labix.org/gopkg.in) to provide versioned imports.
+  * You may use [dep](https://github.com/golang/dep) to manage dependencyes if needed. If you do, do not commit `vendor` directory.
+
+## Build System
+
+* If your project requires fetching or building dependencies, implement that in the `dependencies` rule.
+* If binaries can be built, do so in the `packages` rule. Note that [src-d/ci](https://github.com/src-d/ci) provides that rule by default.
+
+## Testing
+
+* Use [testify](https://github.com/stretchr/testify).
+* Use [suites](https://github.com/stretchr/testify#suite-package) to share setup/teardown code among related tests. When testing multiple methods of the same struct, you generally want to start with a suite. Some cases, such as table-driven tests, are better handled with [subtests](https://blog.golang.org/subtests) though.
+
+## Executables
+
+* Put sources for your executable commands in `cmd/<command-name>/main.go`.
+* We use [go-flags](https://github.com/jessevdk/go-flags) extensively for CLI options parsing.
+
+## Error handling
+
+* Use [src-d/go-errors](https://github.com/src-d/go-errors).
+
+## Logging
+
+* Use logging in your applications.
+* Use [logrus](https://github.com/sirupsen/logrus) for logging.
+* Use [fields](https://github.com/sirupsen/logrus#fields) instead of string formatting in the log message.
+
+## Other libraries
+
+* We use protobuf extensively and [gogo](https://github.com/gogo/protobuf) is our preferred library.
+
+## Best Practices
+
+* Use `gofmt`, `goimports`, `go vet`.
+* Follow community well-established best practices: [Effective Go](https://golang.org/doc/effective_go.html) and [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments).
+* Sourcerers use additional linters such as [golint](https://github.com/golang/lint) and [megacheck](https://github.com/dominikh/go-tools/tree/master/cmd/megacheck).
+
```

Comment:
Haven't seen this in any document, but I think it's a good general rule to have all usages of WaitGroup in one function, if it does both Add and Wait.

Correct:
```golang
func run(){
   var wg sync.WaitGroup

   wg.Add(1)
   go func(){
      defer wg.Done()
      Do()
   }()

   wg.Wait()
}
```

Wrong:
```golang
func run(){
   var wg sync.WaitGroup

   wg.Add(1)
   go Do(&wg)

   wg.Wait()
}

func Do(wg *sync.WaitGroup) {
   defer wg.Done()
   // ...
}
```
or any other variation of passing `wg` to Do (even as a struct field).

This helps to localize syncronization into a signle function and it's easier to review.

Reply:
Since this is something we're not doing consciously today, maybe this can be your first PR to the Go guide once it gets merged. This way it can receive proper attention and discussion on its own.